src/machine-c-api.h

```diff
@@ -82,1894 +83,544 @@ typedef enum { // NOLINT(modernize-use-using)
 /// \param val Receives value of the register.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_read_f(const cm_machine *m, int i, uint64_t *val, char **err_msg);
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_read_f(const cm_machine *m, int i, uint64_t *val);
 
 /// \brief Writes the value of a floating-point register.
-/// \param m Pointer to valid machine instance
+/// \param m Pointer to a valid machine instance.
 /// \param i Register index. Between 0 and F_REG_COUNT-1, inclusive.
 /// \param val New register value.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_write_f(cm_machine *m, int i, uint64_t val, char **err_msg);
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_write_f(cm_machine *m, int i, uint64_t val);
 
-/// \brief Gets the address of a floating-point register.
-/// \param i Register index. Between 0 and F_REG_COUNT-1, inclusive.
-/// \returns Address of the specified register
-CM_API uint64_t cm_get_f_address(int i);
+/// \brief Read the value of any CSR.
+/// \param m Pointer to a valid machine instance.
+/// \param csr CSR to read.
+/// \param val Receives value read from the CSR.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_read_csr(const cm_machine *m, CM_CSR csr, uint64_t *val);
+
+/// \brief Write the value of any CSR.
+/// \param m Pointer to a valid machine instance.
+/// \param csr CSR to write.
+/// \param val Value to write.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_write_csr(cm_machine *m, CM_CSR csr, uint64_t val);
 
 /// \brief Reads the value of the pc register.
-/// \param m Pointer to valid machine instance
+/// \param m Pointer to a valid machine instance.
 /// \param val Receives the value of the register
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_read_pc(const cm_machine *m, uint64_t *val, char **err_msg);
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_read_pc(const cm_machine *m, uint64_t *val);
```

Comment:
This is not very common to use, right? Except for the debugger perhaps? Shouldn't we remove it? You can always use read_csr, or even write your own convenience function for it. Same for `write_pc`.

Reply:
Ok, will remove.
So we implement only for `mcycle`/`uarch_cycle`/`uarch_halt_flag` ? Or should also remove the function for `uarch_halt_flag` ?

Reply:
I think Dave will the halt flag frequently, might as well keep it.

Reply:
See comments on `htif_tohost` and `htif_fromhost` and `iflags_*.*`. I think we should keep them as is for now. But do a radical change later on, where we'd get rid of all `iflags` functions, and keep only `read_htif_tohost` and `write_htif_fromhost`.

Reply:
> ut do a radical change later on, where we'd get rid of all iflags functions
Sure, I will avoid changing how `iflags` works.